### PR TITLE
Bug 743098 - migration before bookmarks and history

### PR DESCRIPTION
**Pull request by ncalexan**
Due to Fennec passwords CP crashiness, hasn't been tested on device.

This disables the engine if the Fennec dbs aren't marked as migrated.

### CODE REVIEW

File: src/main/java/org/mozilla/gecko/sync/repositories/android/FennecControlHelper.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
+package org.mozilla.gecko.sync.repositories.android;
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/repositories/android/FennecControlHelper.java:5]: License.

File: src/main/java/org/mozilla/gecko/sync/repositories/android/FennecControlHelper.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    Cursor cursor = null;
+    try {
+      cursor = queryHelper.safeQuery(providerClient, ".isColumnMigrated(" + column + ")", columns, null, null, null);
+      if (!cursor.moveToFirst()) {
+        return false;
+      }
+      int value = RepoUtils.getIntFromCursor(cursor, column);
+      return value > 0;
+    } finally {
+      if (cursor != null) {
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/repositories/android/FennecControlHelper.java]: If you move the `cursor = queryHelper…` outside the `try`, you can rely on the behavior of `safeQuery` and your `throws` clause and get rid of this condition.

File: src/main/java/org/mozilla/gecko/sync/stage/AndroidBrowserBookmarksServerSyncStage.java [hunk 3]
```diff
@@ -80,4 +84,31 @@ protected Repository getLocalRepository() {
...
+
+  protected boolean areBookmarksMigrated() {
+    // Avoid colliding with a migrating Fennec.
+    if (session == null || session.getContext() == null) {
+      return false;
+    }
+    FennecControlHelper control = null;
+    boolean migrated = false;
+    try {
+      control = new FennecControlHelper(session.getContext());
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/stage/AndroidBrowserBookmarksServerSyncStage.java]: Lift this out of the `try`.

File: src/main/java/org/mozilla/gecko/sync/stage/AndroidBrowserHistoryServerSyncStage.java [hunk 4]
```diff
@@ -80,4 +84,31 @@ protected Repository getRemoteRepository() throws URISyntaxException {
...
+  protected boolean isHistoryMigrated() {
+    // Avoid colliding with a migrating Fennec.
+    if (session == null || session.getContext() == null) {
+      return false;
+    }
+    FennecControlHelper control = null;
+    boolean migrated = false;
+    try {
+      control = new FennecControlHelper(session.getContext());
+      migrated = control.isHistoryMigrated();
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/stage/AndroidBrowserHistoryServerSyncStage.java]: So why not embed all of this safety in `isFooMigrated()`, and add a static version, so you can just do:

```
boolean migrated = FennecControlHelper.isHistoryMigrated(session.getContext());
if (!migrated) {
  // Log
}
return migrated;
```

where

```
public static boolean isHistoryMigrated(Context context) {
  FennecControlHelper h = new FennecControlHelper(context);
  try {
    …
  } finally {
    h.releaseProviders();
  }
}
```

It's the same between both content providers, and you just release…

File: test/src/org/mozilla/gecko/sync/repositories/android/test/TestFennecControlHelper.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+public class TestFennecControlHelper extends AndroidSyncTestCase {
+  public void testIsMigrated() throws NoContentProviderException, NullCursorException, RemoteException {
+    FennecControlHelper control = new FennecControlHelper(getApplicationContext());
+    assertTrue(control.areBookmarksMigrated());
+    // Might take a few shots, but we should be able to migrate if we haven't already.
+    int i = 15;
+    while (!control.isHistoryMigrated()) {
+      assertTrue(i-- > 0);
+    }
```
Comment (rnewman) [inline test/src/org/mozilla/gecko/sync/repositories/android/test/TestFennecControlHelper.java]: Print some log output before this warning that this test will fail if migration takes time!

**Comment by ncalexan** [comment]
I implemented static access very slightly differently than you suggested, but I think it gets the same idea.  We could remove the instance altogether; I was looking ahead to being able to query more things (especially profile data) and assumed we would want to ask for lots of things in a single sync, so a stateful object seemed right.  Let me know if we'd rather go static.

**Pull request merged**